### PR TITLE
WebsocketProvider improvements

### LINKED ISSUE #2711: WebsocketProvider event emitter problem
### Description 

`WebsocketProvider.on 'connect'` event stop occur after first time and does not fired after next socket connect, subscriptions not resubscribed too.

#### Expected behavior

provider = new Web3.providers.WebsocketProvider(wsUrl, options)
onConnect = () => console.log('connected')
provider.on('connect', onConnect)

onConnect callback should be called after each websocket reconnect

#### Actual behavior

onConnect called only once on first connect

#### Steps to reproduce the behavior

Link to example repository: [https://github.com/boomfly/web3-provider-events-example](https://github.com/boomfly/web3-provider-events-example)

#### Error Logs 

[Ethereum] provider "connect" event
pending tx 0xb8c7032fd7a05d7606541deec57940b6680dd52409bdb6125e6eb121dc7e4e3b
pending tx 0x8bdb5473779f7aa37e0bf2eacfa2d017280ba16166bbd24a7dfb4e80ca8b4f45
pending tx 0x0b64b3841fb115bfa25c77072d2402fd9cdb61f20eebc2561354808902a5c74e
pending tx 0x0f9defb78956072193a1fee2b53d730513f22b4ee673ff90b90e0446e3afc67d
pending tx 0x3e3e5a3407b4f38aff3b649b79ac0b57ebd49e637902f67085b3dff01c01aad0
pending tx 0x88d5ed656b1ae27eec4846cca58dff1eb2da87982d4fff5aa105d2051d214999
pending tx 0xefd02a0d9b95e91fa86f4b8ab10b8a0aef52cb43f36691a363da54ab14253b1a
pending tx 0xb347153c82b73a74b20135e3147e82bb2a04005cb1f6b639a6799bb43666ad7a

##### after network disable

Error while cheking connecion Error: Connection error: Timeout exceeded
Check connection connected false, isConnecting false, latestBlock undefined

Error while cheking connecion Error: Connection error: Connection is not open on send()
Check connection connected false, isConnecting false, latestBlock undefined
[Ethereum] provider "error" event { error:
_Event {
 type: 'error',
 isTrusted: false,
 _yaeti: true,
 target:
  W3CWebSocket {
    _listeners: [Object],
    addEventListener: [Function: _addEventListener],
    removeEventListener: [Function: _removeEventListener],
    dispatchEvent: [Function: _dispatchEvent],
    _url: 'wss://ropsten.infura.io/ws',
    _readyState: 3,
    _protocol: undefined,
    _extensions: '',
    _bufferedAmount: 0,
    _binaryType: 'arraybuffer',
    _connection: undefined,
    _client: [WebSocketClient] },
 cancelable: true,
 stopImmediatePropagation: [Function] } }
 Error while cheking connecion Error: Connection error: Connection is not open on send()
 Check connection connected false, isConnecting false, latestBlock undefined

##### and enable network again onConnect not fired and subscriptions not resubscribed, but socket connected

Error while cheking connecion Error: Connection error: Connection is not open on send()
Check connection connected false, isConnecting false, latestBlock undefined
Error while cheking connecion Error: Connection error: Connection is not open on send()
Check connection connected false, isConnecting false, latestBlock undefined
Check connection connected true, isConnecting false, latestBlock 5435004
Check connection connected true, isConnecting false, latestBlock 5435005

### Versions 
* web3.js: 1.0.0-beta.52
* nodejs: 8.15.1 / 10.15.3
* browser:
* ethereum node: wss://ropsten.infura.io/ws or wss://mainnet.infura.io/ws

### PR DESCRIPTION
**Pull request by nivida**
## Description

Fixes #2711 #2851

## Type of change

- [x] Bug fix 

## Checklist:

- [x] I have selected the correct base branch.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no warnings.
- [x] I have updated or added types for all modules I've changed
- [x] Any dependent changes have been merged and published in downstream modules.
- [x] I ran ```npm run test``` in the root folder with success and extended the tests if necessary.
- [x] I ran ```npm run build``` in the root folder and tested it in the browser and with node.
- [x] I ran ```npm run dtslint``` in the root folder and tested that all my types are correct
- [x] I have tested my code on an ethereum test network.

### CODE REVIEW

**Review by boomfly — COMMENTED** [review]
AbstractSocketProvider.js 169 line:
this.removeAllListeners();
this line removes all app defined provider listeners, so we can't receive 'connect', 'reconnected', 'error' events any more

**Review by boomfly — COMMENTED** [review]
removed import isFunction in AbstractSubscription so on line 167, 180 undefined exception

**Comment by boomfly** [comment]
All other seems fine. I hope this PR merged. I need it very much

**Comment by nivida** [comment]
@boomfly Yes, I'll fix all of these. I'm currently testing the reconnecting of the socket connection with different edge cases etc. I will tag you for a final review later:)

File: packages/web3-providers/lib/providers/AbstractSocketProvider.js [hunk 1]
```diff
@@ -165,8 +165,6 @@ export default class AbstractSocketProvider extends EventEmitter {
     onError(error) {
         this.emit(this.ERROR, error);
         this.emit(this.SOCKET_ERROR, error);
-        this.removeAllSocketListeners();
```
Comment (boomfly) [inline packages/web3-providers/lib/providers/AbstractSocketProvider.js]: On reconnection new socket created each time so we need to clear socket events

Comment (nivida) [reply packages/web3-providers/lib/providers/AbstractSocketProvider.js]: I'm currently testing the memory consumption etc. and will do the finish and other tests after. Could you please wait until the ``in-progress`` got removed. Thanks.

**Pull request merged**